reference/test/matrix/csr_kernels.cpp

```diff
@@ -1284,6 +1310,439 @@ TYPED_TEST(Csr, NonSquareMtxIsTransposable)
+    mtx->read(data);
+    return mtx;
+}
+
+
+template <typename ValueType, typename IndexType, typename Permutation>
+std::unique_ptr<gko::matrix::Csr<ValueType, IndexType>> ref_permute(
+    gko::matrix::Csr<ValueType, IndexType>* input, Permutation* permutation,
+    gko::matrix::permute_mode mode)
+{
+    using gko::matrix::permute_mode;
+    using Csr = gko::matrix::Csr<ValueType, IndexType>;
+    auto result = input->clone();
+    auto permutation_csr = csr_from_permutation<ValueType>(
+        permutation, (mode & permute_mode::inverse) == permute_mode::inverse);
+    if ((mode & permute_mode::rows) == permute_mode::rows) {
+        // compute P * A
+        permutation_csr->apply(input, result);
+    }
+    if ((mode & permute_mode::columns) == permute_mode::columns) {
+        // compute A * P^T = (P * A^T)^T
+        auto tmp = result->transpose();
+        auto tmp2 = gko::as<Csr>(tmp->clone());
+        permutation_csr->apply(tmp, tmp2);
+        result = gko::as<Csr>(tmp2->transpose());
+    }
+    return result;
+}
+
+
+template <typename ValueType, typename IndexType, typename Permutation>
+std::unique_ptr<gko::matrix::Csr<ValueType, IndexType>> ref_permute(
+    gko::matrix::Csr<ValueType, IndexType>* input, Permutation* row_permutation,
+    Permutation* col_permutation, bool invert)
+{
+    using gko::matrix::permute_mode;
+    using Csr = gko::matrix::Csr<ValueType, IndexType>;
+    auto result = input->clone();
+    auto row_permutation_csr =
+        csr_from_permutation<ValueType>(row_permutation, invert);
+    auto col_permutation_csr =
+        csr_from_permutation<ValueType>(col_permutation, invert);
+    row_permutation_csr->apply(input, result);
+    auto tmp = result->transpose();
+    auto tmp2 = gko::as<Csr>(tmp->clone());
+    col_permutation_csr->apply(tmp, tmp2);
+    return gko::as<Csr>(tmp2->transpose());
+}
+
+
+TYPED_TEST(Csr, Permute)
+{
+    using gko::matrix::permute_mode;
+
+    for (auto mode :
+         {permute_mode::none, permute_mode::rows, permute_mode::columns,
+          permute_mode::symmetric, permute_mode::inverse_rows,
+          permute_mode::inverse_columns, permute_mode::inverse_symmetric}) {
+        SCOPED_TRACE(mode);
+
+        auto permuted = this->mtx3_sorted->permute(this->perm3, mode);
+        auto ref_permuted =
+            ref_permute(this->mtx3_sorted.get(), this->perm3.get(), mode);
+
+        GKO_ASSERT_MTX_NEAR(permuted, ref_permuted, 0.0);
+        GKO_ASSERT_MTX_EQ_SPARSITY(permuted, ref_permuted);
+        ASSERT_TRUE(permuted->is_sorted_by_column_index());
+    }
+}
+
+
+TYPED_TEST(Csr, PermuteRoundtrip)
+{
+    using gko::matrix::permute_mode;
+
+    for (auto mode :
+         {permute_mode::rows, permute_mode::columns, permute_mode::symmetric}) {
+        SCOPED_TRACE(mode);
+
+        auto permuted =
+            this->mtx3_sorted->permute(this->perm3, mode)
+                ->permute(this->perm3, mode | permute_mode::inverse);
```

Comment:
also perm3_rev for another round trip

Reply:
I added more modes instead to check the inverse case, that should be equivalent?

Reply:
With PermuteInverted, I think so